### PR TITLE
Implement time_none

### PR DESCRIPTION
**Pull request by uramirez8707**
**Description**
1. Implements the "time_none" reduction method
2. Update the test_time_none.sh to test with the new diag manager
3. Fixes a bug related to the subaxis
4. Added the compute indices for the subaxis (this is needed when doing the math)
5. Replace the pointer remapping with a regular copy because we can't guarantee that the data send in is contiguous and that will cause undefined behavior 

~There seems to be an issue with the checks in CI. I have not been able to reproduce the error outside of the CI. The checks are currently commented out.~

Fixes # (issue)

**How Has This Been Tested?**
CI, including new tests!

**Checklist:**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules
- [x] New check tests, if applicable, are included
- [x] `make distcheck` passes

### CODE REVIEW

**Comment by thomas-robinson** [comment]
1.  What is the issue with the CI? 
2. Why can't we guarantee that the incoming data is contiguous?  I'd really like to avoid using copies as much as possible.

File: diag_manager/diag_manager.F90 [hunk 1]
```diff
@@ -1676,9 +1676,9 @@ LOGICAL FUNCTION diag_send_data(diag_field_id, field, time, is_in, js_in, ks_in,
     CHARACTER(len=128) :: error_string, error_string1
 
     REAL, ALLOCATABLE, DIMENSION(:,:,:) :: field_out !< Local copy of field
-    class(*), pointer, dimension(:,:,:,:) :: field_remap !< 4d remapped pointer
-    logical,  pointer, dimension(:,:,:,:) :: mask_remap  !< 4d remapped pointer
-    class(*), pointer, dimension(:,:,:,:) :: rmask_remap !< 4d remapped pointer
+    class(*), allocatable, dimension(:,:,:,:) :: field_remap !< 4d remapped pointer
+    logical,  allocatable, dimension(:,:,:,:) :: mask_remap  !< 4d remapped pointer
+    class(*), allocatable, dimension(:,:,:,:) :: rmask_remap !< 4d remapped pointer
```
Comment (thomas-robinson) [inline diag_manager/diag_manager.F90:1681]: None of these are pointers anymore

File: diag_manager/fms_diag_bbox.F90 [hunk 2]
```diff
@@ -202,11 +286,68 @@ SUBROUTINE update_bounds(this, lower_i, upper_i, lower_j, upper_j, lower_k, uppe
...
+      class(*),                     intent(in)    :: field_data(:,:,:,:) !< Field data
+      INTEGER,                      INTENT(in)    :: lower_i             !< Lower i bound.
+      INTEGER,                      INTENT(in)    :: upper_i             !< Upper i bound.
+      INTEGER,                      INTENT(in)    :: lower_j             !< Lower j bound.
+      INTEGER,                      INTENT(in)    :: upper_j             !< Upper j bound.
+      INTEGER,                      INTENT(in)    :: lower_k             !< Lower k bound.
+      INTEGER,                      INTENT(in)    :: upper_k             !< Upper k bound.
+      LOGICAL,                      INTENT(in)    :: has_halos           !< .true. if the field has halos
+
+      character(len=150) :: error_msg
```
Comment (thomas-robinson) [inline diag_manager/fms_diag_bbox.F90:303]: Should have a doxygen comment

File: diag_manager/fms_diag_bbox.F90 [hunk 3]
```diff
@@ -30,7 +30,8 @@
 !> @{
 MODULE fms_diag_bbox_mod
 
-   USE fms_mod, ONLY: error_mesg, FATAL, fms_error_handler
+   USE fms_mod, ONLY: error_mesg, FATAL, fms_error_handler, string
+   use mpp_mod
```
Comment (thomas-robinson) [inline diag_manager/fms_diag_bbox.F90:34]: Use an `only` statements.

File: diag_manager/fms_diag_bbox.F90 [hunk 4]
```diff
@@ -81,10 +88,52 @@ MODULE fms_diag_bbox_mod
       procedure :: get_fje
    end type fmsDiagBoundsHalos_type
 
-   public :: recondition_indices
+   public :: recondition_indices, determine_if_block_is_in_region
+
+   integer, parameter :: xdimension = 1
+   integer, parameter :: ydimension = 2
+   integer, parameter :: zdimension = 3
```
Comment (thomas-robinson) [inline diag_manager/fms_diag_bbox.F90:95]: These need doxygen comments

File: diag_manager/fms_diag_bbox.F90 [hunk 5]
```diff
@@ -81,10 +88,52 @@ MODULE fms_diag_bbox_mod
...
+    return
+  endif
+
+  if (block_start > subregion_end   .and. block_end > subregion_end) then
+    determine_if_block_is_in_region = .false.
+    return
+  endif
+
+  determine_if_block_is_in_region = .true.
+end function
```
Comment (thomas-robinson) [inline diag_manager/fms_diag_bbox.F90:135]: Put the function name here

File: diag_manager/fms_diag_bbox.F90 [hunk 6]
```diff
@@ -319,6 +460,48 @@ function recondition_indices(indices, field, is_in, js_in, ks_in, &
    indices%fje = fje
  end function recondition_indices
 
+ !> @brief Rebase the ouput bounds for a given dimension based on the starting and ending indices
```
Comment (thomas-robinson) [inline diag_manager/fms_diag_bbox.F90:463]: Make these comments more descriptive

File: diag_manager/fms_diag_object.F90 [hunk 7]
```diff
@@ -714,22 +727,137 @@ subroutine fms_diag_do_io(this, is_end_of_run)
...
+    reduction_method = field_yaml_ptr%get_var_reduction()
+    select case(reduction_method)
+    case (time_none)
+      error_msg = buffer_ptr%do_time_none_wrapper(field_data, oor_mask, bounds_in, bounds_out)
+      if (trim(error_msg) .ne. "") then
+        return
+      endif
+    case (time_min)
+    case (time_max)
+    case (time_sum)
```
Comment (thomas-robinson) [inline diag_manager/fms_diag_object.F90:851]: Missing mean/average

File: diag_manager/fms_diag_object.F90 [hunk 8]
```diff
@@ -714,22 +727,137 @@ subroutine fms_diag_do_io(this, is_end_of_run)
...
+    case (time_power)
+    case (time_rms)
+    case (time_diurnal)
+    end select
```
Comment (thomas-robinson) [inline diag_manager/fms_diag_object.F90:855]: There should be a default with an error

File: diag_manager/include/fms_diag_reduction_methods_r4.fh [hunk 9]
```diff
@@ -0,0 +1,32 @@
+!* This file is part of the GFDL Flexible Modeling System (FMS).
```
Comment (thomas-robinson) [inline diag_manager/include/fms_diag_reduction_methods_r4.fh:1]: This is missing a line I think...

**Review by thomas-robinson — CHANGES_REQUESTED** [review]

**Comment by uramirez8707** [comment]
The "CI issue" is fixed in https://github.com/NOAA-GFDL/FMS/pull/1347/commits/e0cc678bf9713f73031c36ccc54ae67b14bc5523. The CI was failing with `--enable-mixed-mode`, it was not getting the subregion indices correctly.  I turned the checks back on.

File: diag_manager/diag_manager.F90 [hunk 10]
```diff
@@ -4480,6 +4480,29 @@ SUBROUTINE diag_field_add_cell_measures(diag_field_id, area, volume)
...
+      type is (real(kind=r8_kind))
+        data_out(:,:,:,1) = data_in
+      end select
+    type is (real(kind=r4_kind))
+      allocate(real(kind=r4_kind) :: data_out(1:size(data_in,1), 1:size(data_in,2), 1:size(data_in,3), 1))
+      select type (data_out)
+      type is (real(kind=r4_kind))
+        data_out(:,:,:,1) = data_in
+      end select
+    end select
```
Comment (rem1776) [inline diag_manager/diag_manager.F90:4503]: I think there should be some type of error check for types here, otherwise it could silently fail.

Comment (uramirez8707) [reply diag_manager/diag_manager.F90:4517]: https://github.com/NOAA-GFDL/FMS/pull/1347/commits/be486f72f223f6c14379557ea8abaa55c761929e

**Review by rem1776 — APPROVED** [review]

**Pull request merged**